### PR TITLE
lnd: optionally serve pprof on localhost for better security

### PR DESCRIPTION
**Pull request by Crypt-iQ**
Otherwise, if the port is connectable, anybody can connect and see debugging information.

See: https://mmcloughlin.com/posts/your-pprof-is-showing

### CODE REVIEW

**Review by Roasbeef — APPROVED** [review]
Solid change!

Switching it like this may break some existing set ups, but I think it's fine as long as we include it in a major release and point out the changes in the release notes clearly. 

LGTM 🍬

**Comment by joostjager** [comment]
What is your suggestion for accessing the profiler remotely after this change? 

Alternatively the `--profile` option could take `ip:port` similar to other listening configuration options. Then users still have the possibility to access remotely.

**Comment by joostjager** [comment]
>What setup prevents somebody from ssh + accessing the profiler?

That that person doesn't have ssh access because of the access to the coins that comes with it. One could argue that ideally nobody has ssh access for security reasons.

**Comment by Crypt-iQ** [comment]
> > What setup prevents somebody from ssh + accessing the profiler?
> 
> That that person doesn't have ssh access because of the access to the coins that comes with it. One could argue that ideally nobody has ssh access for security reasons.

How is the remote lnd being controlled?

**Comment by tanuck** [comment]
I think node operators should at least be given the option to make this decision for themselves. :100: in favour of sensible defaults that don't expose profiling data.

Giving SSH access just to fetch profiling data is a bit much IMO.

**Comment by joostjager** [comment]
>How is the remote lnd being controlled?

Through grpc with restricted macaroons or via other applications that offer a higher level interface to the operator.

**Comment by cfromknecht** [comment]
> Through grpc with restricted macaroons or via other applications that offer a higher level interface to the operator.

It would be an interesting feature to somehow expose the profiling via our API so that it can be accessed securely.

In the meantime I think it makes sense to remove listening on public interfaces as the _default_ behavior, but still allow users to do so if they need to.

**Comment by Crypt-iQ** [comment]
> > Through grpc with restricted macaroons or via other applications that offer a higher level interface to the operator.
> 
> It would be an interesting feature to somehow expose the profiling via our API so that it can be accessed securely.
> 
> In the meantime I think it makes sense to remove listening on public interfaces as the _default_ behavior, but still allow users to do so if they need to.

I was thinking the same thing re: gRPC to get profiling information.  I can change the public interface listening to non-default, with a warning/documentation that it reduces security.

**Review by Roasbeef — CHANGES_REQUESTED** [review]
Echoing @guggero's comment for the same change in a related private repo: 

> 👍 for making this configurable. How about this: We change the config.Profile to a string and parse it. If it's just a number (e.g. to not break existing users that have the flag set), we use 127.0.0.1 as the host. If it's a host:port pair, use that directly.

**Comment by Crypt-iQ** [comment]
> Echoing @guggero's comment for the same change in a related private repo:
> 
> > 👍 for making this configurable. How about this: We change the config.Profile to a string and parse it. If it's just a number (e.g. to not break existing users that have the flag set), we use 127.0.0.1 as the host. If it's a host:port pair, use that directly.

Will do this

**Comment by Crypt-iQ** [comment]
Updated so that `--profile` can be set to either a port or host:port.

File: lnd.go [hunk 1]
```diff
@@ -227,11 +227,10 @@ func Main(cfg *Config, lisCfg ListenerCfg, interceptor signal.Interceptor) error
 	// Enable http profiling server if requested.
 	if cfg.Profile != "" {
 		go func() {
-			listenAddr := net.JoinHostPort("", cfg.Profile)
 			profileRedirect := http.RedirectHandler("/debug/pprof",
 				http.StatusSeeOther)
 			http.Handle("/", profileRedirect)
-			fmt.Println(http.ListenAndServe(listenAddr, nil))
+			fmt.Println(http.ListenAndServe(cfg.Profile, nil))
```
Comment (Roasbeef) [inline lnd.go:233]: We should add a log here to show the final interface we end up listening on. May be useful in debugging why things aren't showing up.

**Review by Roasbeef — APPROVED** [review]
LGMT 🛵

Verified (using `lsof`) it's possible to listen on both local host w/ a port as well as a custom interface: 
```
lnd-debug 43224 roasbeef   12u     IPv4 0x999f56ad153ba635        0t0                 TCP localhost:7050 (LISTEN)
lnd-debug 42508 roasbeef   13u     IPv6 0x999f56ad155729c5        0t0                 TCP *:7050 (LISTEN)
```
First is `--profile=0.0.0.0:7050`, second is `--profile=7050`.

File: lnd.go (hunk 1 above)
Comment (Crypt-iQ) [reply lnd.go:234]: Added log

**Pull request merged**